src/modules/fancyzones/lib/FancyZones.cpp

```diff
@@ -961,31 +966,68 @@ void FancyZones::HandleVirtualDesktopUpdates(HANDLE fancyZonesDestroyedEvent) no
-        std::unordered_map<GUID, bool> temp;
-        temp.reserve(bufferCapacity / guidSize);
-        for (size_t i = 0; i < bufferCapacity; i += guidSize)
-        {
-            GUID* guid = reinterpret_cast<GUID*>(buffer.get() + i);
-            temp[*guid] = true;
-        }
-        std::unique_lock writeLock(m_lock);
-        for (auto it = begin(m_virtualDesktopIds); it != end(m_virtualDesktopIds);)
+        RegisterVirtualDesktopUpdates(buffer, bufferCapacity);
+    }
+}
+
+void FancyZones::RegisterVirtualDesktopUpdates(std::unique_ptr<BYTE[]>& buffer, int bufferCapacity) noexcept
+{
+    const int guidSize = sizeof(GUID);
+    std::unordered_set<GUID> temp;
+    temp.reserve(bufferCapacity / guidSize);
+    for (size_t i = 0; i < bufferCapacity; i += guidSize)
+    {
+        GUID* guid = reinterpret_cast<GUID*>(buffer.get() + i);
+        temp.insert(*guid);
+    }
+    std::unique_lock writeLock(m_lock);
+    for (auto it = begin(m_processedWorkAreas); it != end(m_processedWorkAreas);)
+    {
+        auto iter = temp.find(it->first);
+        if (iter == temp.end())
         {
-            auto iter = temp.find(it->first);
-            if (iter == temp.end())
-            {
-                it = m_virtualDesktopIds.erase(it); // virtual desktop closed, remove it from map
-            }
-            else
+            // clean up data related to virtual desktop
+            wil::unique_cotaskmem_string virtualDesktopId;
+            if (SUCCEEDED_LOG(StringFromCLSID(it->first, &virtualDesktopId)))
             {
-                temp.erase(it->first); // virtual desktop already in map, skip it
-                ++it;
+                JSONHelpers::FancyZonesDataInstance().RemoveDevicesByVirtualDesktopId(virtualDesktopId.get());
             }
+            it = m_processedWorkAreas.erase(it); // virtual desktop closed, remove it from map
         }
-        // register new virtual desktops, if any
-        m_virtualDesktopIds.insert(begin(temp), end(temp));
+        else
+        {
+            temp.erase(it->first); // virtual desktop already in map, skip it
+            ++it;
+        }
+    }
+    // register new virtual desktops, if any
+    for (const auto& id : temp)
+    {
+        m_processedWorkAreas[id] = std::vector<HMONITOR>();
+    }
+}
+
+void FancyZones::RegisterNewWorkArea(GUID virtualDesktopId, HMONITOR monitor) noexcept
+{
+    if (!m_processedWorkAreas.contains(virtualDesktopId))
+    {
+        m_processedWorkAreas[virtualDesktopId] = { monitor };
     }
+    else
+    {
+        m_processedWorkAreas[virtualDesktopId].push_back(monitor);
+    }
+}
+
+bool FancyZones::IsNewWorkArea(GUID virtualDesktopId, HMONITOR monitor) noexcept
+{
+    auto it = m_processedWorkAreas.find(virtualDesktopId);
+    if (it != m_processedWorkAreas.end())
+    {
+        // virtual desktop exists, check if it's processed on given monitor
+        return std::find(it->second.begin(), it->second.end(), monitor) == it->second.end();

```

Comment:
Maybe:
```
return std::none_of(it->second.begin(), it->second.end(), [monitor](HMONITOR m) { return m == monitor; });
```

Reply:
I think not having an additional negation doesn't reduce readability 🙂